File: packages/gavio-java/gavio-core/src/test/java/io/gavio/inspector/InspectorApiVectorsTest.java
package io.gavio.inspector;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import io.gavio.Gateway;
import io.gavio.GavioRequest;
import io.gavio.json.Json;
import io.gavio.providers.MockProvider;
import io.gavio.types.Message;
import io.gavio.types.Provider;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

/**
 * Runs the shared v0.7.0 Inspector API cases from
 * {@code //test-vectors/inspector/api-cases.json} against the Java SDK —
 * DAG assembly (F-OBS-10) and replay-mode gating (F-DX-11). Same file the
 * Python and JavaScript suites consume.
 */
class InspectorApiVectorsTest {

    private static final HttpClient CLIENT = HttpClient.newHttpClient();

    /** Walk up from the working dir to find the repo's test-vectors directory. */
    private static Path vectorsFile() {
        Path dir = Path.of("").toAbsolutePath();
        while (dir != null) {
            Path candidate = dir.resolve("test-vectors/inspector/api-cases.json");
            if (Files.isRegularFile(candidate)) {
                return candidate;
            }
            dir = dir.getParent();
        }
        throw new IllegalStateException("could not locate test-vectors/inspector from working dir");
    }

    private static Map<String, Object> vectors() throws IOException {
        return Json.parseObject(Files.readString(vectorsFile()));
    }

    @SuppressWarnings("unchecked")
    static Stream<Map<String, Object>> dagCases() throws IOException {
        return ((List<Object>) vectors().get("dagCases")).stream()
                .map(c -> (Map<String, Object>) c);
    }

    @SuppressWarnings("unchecked")
    static Stream<Map<String, Object>> replayGatingCases() throws IOException {
        return ((List<Object>) vectors().get("replayGating")).stream()
                .map(c -> (Map<String, Object>) c);
    }

    @SuppressWarnings("unchecked")
    @ParameterizedTest(name = "{0}")
    @MethodSource("dagCases")
    void dagVectors(Map<String, Object> dagCase) {
        List<Map<String, Object>> summaries =
                (List<Map<String, Object>>) (List<?>) dagCase.get("summaries");
        Map<String, Object> dag = InspectorAnalytics.buildDag(
                summaries, (String) dagCase.get("root"), (String) dagCase.get("sessionId"));
        assertNotNull(dag, (String) dagCase.get("id"));

        Map<String, Object> expected = (Map<String, Object>) dagCase.get("expected");
        List<Map<String, Object>> nodes = (List<Map<String, Object>>) (List<?>) dag.get("nodes");
        List<?> edges = (List<?>) dag.get("edges");
        assertEquals(((Number) expected.get("nodes")).intValue(), nodes.size());
        assertEquals(((Number) expected.get("edges")).intValue(), edges.size());

        String rootId = (String) dagCase.get("root");
        if (rootId == null) {
            rootId = nodes.stream()
                    .filter(n -> n.get("parentTraceId") == null)
                    .findFirst()
                    .orElseThrow()
                    .get("traceId")
                    .toString();
        }
        final String root = rootId;
        Map<String, Object> rootNode = nodes.stream()
                .filter(n -> root.equals(n.get("traceId")))
                .findFirst()
                .orElseThrow();
        Map<String, Object> subtree = (Map<String, Object>) rootNode.get("subtree");
        Map<String, Object> expectedSubtree = (Map<String, Object>) expected.get("rootSubtree");
        for (Map.Entry<String, Object> entry : expectedSubtree.entrySet()) {
            double want = ((Number) entry.getValue()).doubleValue();
            double got = ((Number) subtree.get(entry.getKey())).doubleValue();
            assertEquals(want, got, 1e-8, entry.getKey());
        }
    }

    @ParameterizedTest(name = "mode={0}")
    @MethodSource("replayGatingCases")
    void replayGatingVectors(Map<String, Object> gating) throws Exception {
        CaptureMode mode = CaptureMode.valueOf(((String) gating.get("mode")).toUpperCase());
        int expectedStatus = ((Number) gating.get("expectedStatus")).intValue();

        Gateway gateway = Gateway.builder()
                .adapter(new MockProvider())
                .model("mock")
                .inspect(InspectorConfig.builder()
                        .enabled(true)
                        .port(0)
                        .mode(mode)
                        .unsafeContentCaptureAck(true)
                        .build())
                .build();
        try {
            String traceId = gateway.complete(GavioRequest.builder()
                            .messages(List.of(Message.of("user", "gate me")))
                            .model("mock")
                            .provider(Provider.MOCK)
                            .build())
                    .join()
                    .traceId();
            String base = "http://127.0.0.1:" + gateway.inspector().port();
            HttpRequest request = HttpRequest.newBuilder(URI.create(base + "/api/replay"))
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(
                            Json.write(Map.of("traceId", traceId))))
                    .build();
            HttpResponse<String> response =
                    CLIENT.send(request, HttpResponse.BodyHandlers.ofString());
            assertEquals(expectedStatus, response.statusCode());
            if (expectedStatus == 200) {
                Map<String, Object> body = Json.parseObject(response.body());
                assertTrue(body.get("traceId") instanceof String);
                assertEquals(traceId, body.get("replayOf"));
            }
        } finally {
            gateway.inspector().stop();
        }
    }
}

File: packages/gavio-js/tests/unit/inspector-api-vectors.test.ts
/**
 * Runs the shared v0.7.0 Inspector API cases from
 * //test-vectors/inspector/api-cases.json against the JS SDK — DAG assembly
 * (F-OBS-10) and replay-mode gating (F-DX-11). Same file the Python and Java
 * suites consume.
 */

import { readFileSync } from 'node:fs'
import { describe, it, expect, afterEach } from 'vitest'
import { Gateway } from '../../src/gateway.js'
import { buildCostReport, buildDag } from '../../src/inspector/analytics.js'
import type { SummaryLike } from '../../src/inspector/analytics.js'
import type { InspectorMode } from '../../src/inspector/index.js'
import { mockProvider } from '../../src/providers/mock.js'

interface DagCase {
  id: string
  summaries: SummaryLike[]
  root?: string
  sessionId?: string
  expected: {
    nodes: number
    edges: number
    rootSubtree: { traces: number; errors: number; costUsd: number; latencyMs: number }
  }
}

interface ReplayGatingCase {
  mode: InspectorMode
  expectedStatus: number
}

const vectors = JSON.parse(
  readFileSync(
    new URL('../../../../test-vectors/inspector/api-cases.json', import.meta.url),
    'utf-8',
  ),
) as { dagCases: DagCase[]; replayGating: ReplayGatingCase[] }

const costVectors = JSON.parse(
  readFileSync(
    new URL('../../../../test-vectors/inspector/cost-report.json', import.meta.url),
    'utf-8',
  ),
) as {
  cases: Array<{
    id: string
    groupBy?: string
    summaries: SummaryLike[]
    expected: Record<string, any>
  }>
}

describe('shared test-vectors — inspector/api-cases.json dagCases', () => {
  for (const dagCase of vectors.dagCases) {
    it(dagCase.id, () => {
      const dag = buildDag(dagCase.summaries, dagCase.root, dagCase.sessionId)
      expect(dag).not.toBeNull()
      expect(dag!.nodes).toHaveLength(dagCase.expected.nodes)
      expect(dag!.edges).toHaveLength(dagCase.expected.edges)
      const rootId =
        dagCase.root ?? dag!.nodes.find((n) => n.parentTraceId === null)!.traceId
      const rootNode = dag!.nodes.find((n) => n.traceId === rootId)!
      const expected = dagCase.expected.rootSubtree
      expect(rootNode.subtree.traces).toBe(expected.traces)
      expect(rootNode.subtree.errors).toBe(expected.errors)
      expect(rootNode.subtree.costUsd).toBeCloseTo(expected.costUsd, 8)
      expect(rootNode.subtree.latencyMs).toBe(expected.latencyMs)
    })
  }
})

describe('shared test-vectors — inspector/api-cases.json replayGating', () => {
  const gateways: Gateway[] = []

  afterEach(async () => {
    for (const gw of gateways.splice(0)) await gw.inspector?.close()
  })

  for (const gating of vectors.replayGating) {
    it(`mode=${gating.mode} → ${gating.expectedStatus}`, async () => {
      const gw = new Gateway({
        model: 'mock',
        inspect: { mode: gating.mode, port: 0, unsafeContentCaptureAck: true },
      }).withAdapter(mockProvider())
      gateways.push(gw)
      const port = await gw.inspector!.ready()
      const base = `http://127.0.0.1:${port}`

      const original = await gw.complete({ messages: [{ role: 'user', content: 'gate me' }] })
      const response = await fetch(`${base}/api/replay`, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify({ traceId: original.traceId }),
      })
      expect(response.status).toBe(gating.expectedStatus)
      if (gating.expectedStatus === 200) {
        const body = (await response.json()) as { traceId: string; replayOf: string }
        expect(body.traceId).toBeTruthy()
        expect(body.replayOf).toBe(original.traceId)
      }
    })
  }
})

describe('shared test-vectors — inspector/cost-report.json', () => {
  for (const costCase of costVectors.cases) {
    it(costCase.id, () => {
      const report = buildCostReport(costCase.summaries, costCase.groupBy)
      const expected = costCase.expected
      expect(report.total.requests).toBe(expected.total.requests)
      expect(report.total.errors).toBe(expected.total.errors)
      expect(report.total.tokens).toEqual(expected.total.tokens)
      expect(report.total.costUsd).toBeCloseTo(expected.total.costUsd, 8)
      expect(report.total.averageCostUsd).toBeCloseTo(expected.total.averageCostUsd, 8)
      expect(report.total.cacheHits).toBe(expected.total.cacheHits)
      expect(report.total.retryCount).toBe(expected.total.retryCount)
      expect(report.total.retryOverheadUsd).toBeCloseTo(expected.total.retryOverheadUsd, 8)
      expect(report.total.cacheSavingsUsd).toBeCloseTo(expected.total.cacheSavingsUsd, 8)
      for (const [key, aggregate] of Object.entries(expected.groups)) {
        expect(report.groups![key]!.requests).toBe(aggregate.requests)
        expect(report.groups![key]!.costUsd).toBeCloseTo(aggregate.costUsd, 8)
        if (aggregate.cacheHits !== undefined) {
          expect(report.groups![key]!.cacheHits).toBe(aggregate.cacheHits)
        }
        if (aggregate.errors !== undefined) {
          expect(report.groups![key]!.errors).toBe(aggregate.errors)
        }
      }
      expect(report.topSpend.tenant.slice(0, 2)).toEqual(expected.topSpend.tenant)
      expect(report.topSpend.feature.slice(0, 2)).toEqual(expected.topSpend.feature)
    })
  }
})
